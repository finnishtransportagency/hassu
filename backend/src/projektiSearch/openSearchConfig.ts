export const openSearchConfig = {
  searchDomain: process.env.SEARCH_DOMAIN,
  opensearchYllapitoIndex: "projekti-" + process.env.ENVIRONMENT + "-yllapito",
  opensearchJulkinenIndexPrefix: "projekti-" + process.env.ENVIRONMENT + "-julkinen-",
  opensearchIlmoitustauluSyoteIndex: "projekti-" + process.env.ENVIRONMENT + "-ilmoitustaulusyote",
};
