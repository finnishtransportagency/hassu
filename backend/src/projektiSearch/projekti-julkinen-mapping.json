{
  "properties": {
    "maakunnat": {
      "type": "integer"
    },
    "kunnat": {
      "type": "integer"
    },
    "nimi": {
      "type": "text",
      "fields": {
        "keyword": {
          "type": "keyword",
          "ignore_above": 256
        }
      },
      "analyzer": "sortable"
    },
    "paivitetty": {
      "type": "date"
    },
    "publishTimestamp": {
      "type": "date"
    },
    "projektiTyyppi": {
      "type": "text",
      "fields": {
        "keyword": {
          "type": "keyword",
          "ignore_above": 256
        }
      }
    },
    "vaihe": {
      "type": "text",
      "fields": {
        "keyword": {
          "type": "keyword",
          "ignore_above": 256
        }
      }
    },
    "vaylamuoto": {
      "type": "text",
      "fields": {
        "keyword": {
          "type": "keyword",
          "ignore_above": 256
        }
      }
    },
    "viimeisinJulkaisu": {
      "type": "date"
    }
  }
}
