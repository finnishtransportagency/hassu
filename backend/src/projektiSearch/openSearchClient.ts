import { HttpRequest } from "@aws-sdk/protocol-http";
import { openSearchConfig } from "./openSearchConfig";
import { sendSignedRequest } from "../aws/awsRequest";
import { log } from "../logger";

const domain = openSearchConfig.searchDomain ?? "search-domain-missing";
const type = "_doc";

export async function sendRequest(request: HttpRequest): Promise<unknown> {
  return (await sendSignedRequest(request, "es")).body;
}

export interface SortOrder {
  order: "asc" | "desc";
}

export enum OpenSearchIndexType {
  YLLAPITO = "yllapito",
  JULKINEN = "julkinen",
}

export interface SearchOpts {
  from?: number;
  size?: number;
  sort?: Record<string, SortOrder>[];
  query?: unknown;
  aggs?: unknown;
}

export default class OpenSearchClient {
  private index: string;

  constructor(index: string) {
    this.index = index;
  }

  async putDocument(key: string, document: unknown): Promise<void> {
    const response = (await this.put("/" + type + "/" + key, JSON.stringify(document))) as { result: string };
    if (response["result"] !== "created" && response["result"] !== "updated") {
      log.warn(response);
    }
  }

  async deleteDocument(key: string): Promise<void> {
    await this.delete("/" + type + "/" + key);
  }

  async delete(path: string): Promise<unknown> {
    const request = new HttpRequest({
      headers: {
        host: domain,
      },
      hostname: domain,
      method: "DELETE",
      path: this.index + path,
    });
    return sendRequest(request);
  }

  async deleteIndex(): Promise<unknown> {
    return this.delete("");
  }

  async createIndex(settings: any, mappings: any): Promise<unknown> {
    const body = JSON.stringify({ settings, mappings });
    return this.put("", body);
  }

  async indexExists(): Promise<boolean> {
    const request = new HttpRequest({
      headers: {
        host: domain,
      },
      query: { pretty: "true" },
      hostname: domain,
      method: "HEAD",
      path: this.index,
    });
    try {
      const response = await sendSignedRequest(request, "es");
      return response.statusCode === 200;
    } catch (e) {
      log.info("No index found", e);
    }
    return false;
  }

  async query(query: SearchOpts): Promise<any> {
    const body = JSON.stringify(query);
    log.info("query " + this.index, { query });
    const request = new HttpRequest({
      body,
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      hostname: domain,
      method: "POST",
      path: this.index + "/_search",
    });
    return sendRequest(request);
  }

  async getSettings(): Promise<unknown> {
    return this.get("/_settings");
  }

  async putSettings(settings: string): Promise<unknown> {
    await this.closeIndex();

    try {
      return await this.put("/_settings", settings);
    } finally {
      await this.openIndex();
    }
  }

  async getMapping(): Promise<unknown> {
    const mapping = await this.get("/_mapping");
    return (mapping as any).projekti.mappings;
  }

  async putMapping(mapping: string): Promise<unknown> {
    return this.put("/_mapping", mapping);
  }

  async get(path: string): Promise<any> {
    const request = new HttpRequest({
      headers: {
        host: domain,
      },
      query: { pretty: "true" },
      hostname: domain,
      method: "GET",
      path: this.index + path,
    });
    return sendRequest(request);
  }

  async put(path: string, body: string): Promise<unknown> {
    const request = new HttpRequest({
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      body,
      hostname: domain,
      method: "PUT",
      path: this.index + path,
    });
    return sendRequest(request);
  }

  protected async closeIndex(): Promise<void> {
    log.info("closeIndex " + this.index);
    const request = new HttpRequest({
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      hostname: domain,
      method: "POST",
      path: this.index + "/_close",
    });
    log.info(await sendRequest(request));
  }

  private async openIndex() {
    log.info("openIndex " + this.index);
    const request = new HttpRequest({
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      hostname: domain,
      method: "POST",
      path: this.index + "/_open",
    });
    log.info(sendRequest(request));
  }
}
