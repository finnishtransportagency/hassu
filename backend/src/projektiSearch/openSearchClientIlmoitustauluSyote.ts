import { openSearchConfig } from "./openSearchConfig";
import OpenSearchClient from "./openSearchClient";

class OpenSearchClientIlmoitustauluSyote extends OpenSearchClient {
  constructor() {
    super(openSearchConfig.opensearchIlmoitustauluSyoteIndex);
  }
}

export const openSearchClientIlmoitustauluSyote = new OpenSearchClientIlmoitustauluSyote();
