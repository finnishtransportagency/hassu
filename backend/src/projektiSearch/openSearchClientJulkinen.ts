import { Kieli } from "hassu-common/graphql/apiModel";
import OpenSearchClient from "./openSearchClient";
import { openSearchConfig } from "./openSearchConfig";

class OpenSearchClientJulkinen extends OpenSearchClient {
  constructor(kieli: Kieli) {
    super(openSearchConfig.opensearchJulkinenIndexPrefix + kieli.toLowerCase());
  }
}

export const openSearchClientJulkinen = {
  [Kieli.SUOMI]: new OpenSearchClientJulkinen(Kieli.SUOMI),
  [Kieli.RUOTSI]: new OpenSearchClientJulkinen(Kieli.RUOTSI),
};
