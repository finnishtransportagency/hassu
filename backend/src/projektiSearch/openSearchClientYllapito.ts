import OpenSearchClient from "./openSearchClient";
import { openSearchConfig } from "./openSearchConfig";

class OpenSearchClientYllapito extends OpenSearchClient {
  constructor() {
    super(openSearchConfig.opensearchYllapitoIndex);
  }
}

const openSearchClientYllapito = new OpenSearchClientYllapito();
export default openSearchClientYllapito;
