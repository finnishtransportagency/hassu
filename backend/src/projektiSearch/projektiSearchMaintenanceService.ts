import { openSearchClient } from "./openSearchClient";
import { projektiDatabase } from "../database/projektiDatabase";
import { projektiSearchService } from "./projektiSearchService";
import projektiSettings from "./projekti-settings.json";
import projektiMapping from "./projekti-mapping.json";
import { log } from "../logger";

export type MaintenanceEvent = {
  action: "deleteIndex" | "index";
  startKey?: string;
};

export class ProjektiSearchMaintenanceService {
  async deleteIndex(): Promise<void> {
    log.info(await openSearchClient.deleteIndex());
    log.info(await openSearchClient.put("projekti", "", "{}"));
    log.info(await openSearchClient.putSettings("projekti", JSON.stringify(projektiSettings)));
    log.info(await openSearchClient.putMapping("projekti", JSON.stringify(projektiMapping)));
  }

  async index(event: MaintenanceEvent): Promise<string> {
    const scanResult = await projektiDatabase.scanProjektit(event.startKey);
    for (const projekti of scanResult.projektis) {
      const response = await projektiSearchService.indexProjekti(projekti);
      if (response["result"] !== "created") {
        log.warn(response);
      }
    }
    return scanResult.startKey;
  }
}
