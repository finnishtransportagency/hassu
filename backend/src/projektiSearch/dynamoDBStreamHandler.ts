import { log } from "../logger";
import { DynamoDBRecord, DynamoDBStreamEvent } from "aws-lambda/trigger/dynamodb-stream";
import DynamoDB from "aws-sdk/clients/dynamodb";
import { DBProjekti } from "../database/model/projekti";
import { projektiSearchService } from "./projektiSearchService";
import { setupLambdaMonitoring, setupLambdaMonitoringMetaData } from "../aws/monitoring";
import * as AWSXRay from "aws-xray-sdk-core";
import { MaintenanceEvent, ProjektiSearchMaintenanceService } from "./projektiSearchMaintenanceService";
import { invokeLambda } from "../aws/lambda";
import { Context } from "aws-lambda";

const parse = DynamoDB.Converter.unmarshall;

async function handleUpdate(record: DynamoDBRecord) {
  if (record.dynamodb?.NewImage) {
    const projekti = parse(record.dynamodb.NewImage) as DBProjekti;
    log.info(`${record.eventName}`, { oid: projekti.oid });
    await projektiSearchService.indexProjekti(projekti);
  } else {
    log.error("No DynamoDB record to update");
  }
}

async function handleRemove(record: DynamoDBRecord) {
  if (record.dynamodb?.Keys?.oid.S) {
    const oid: string = record.dynamodb.Keys.oid.S;
    log.info("REMOVE", { oid });
    await projektiSearchService.removeProjekti(oid);
  } else {
    log.error("No DynamoDB key to remove");
  }
}

async function handleManagementAction(action: "deleteIndex" | "index", event: MaintenanceEvent, context: Context) {
  if (action == "deleteIndex") {
    await new ProjektiSearchMaintenanceService().deleteIndex();
  } else if (action == "index") {
    const startKey = await new ProjektiSearchMaintenanceService().index(event);
    if (startKey && startKey !== event.startKey) {
      await invokeLambda(
        context.functionName,
        false,
        JSON.stringify({ action: "index", startKey } as MaintenanceEvent)
      );
    }
  }
}

export const handleDynamoDBEvents = async (
  event: DynamoDBStreamEvent | MaintenanceEvent,
  context: Context
): Promise<void> => {
  const action = (event as MaintenanceEvent).action;
  if (action) {
    return await handleManagementAction(action, event as MaintenanceEvent, context);
  }
  setupLambdaMonitoring();
  if (!(event as DynamoDBStreamEvent).Records) {
    log.warn("No records");
    return;
  }
  return await AWSXRay.captureAsyncFunc("handler", async (subsegment) => {
    try {
      return await (async () => {
        setupLambdaMonitoringMetaData(subsegment);
        for (const record of (event as DynamoDBStreamEvent).Records) {
          switch (record.eventName) {
            case "INSERT":
            case "MODIFY":
              await handleUpdate(record);
              break;
            case "REMOVE":
              await handleRemove(record);
          }
        }
      })();
    } finally {
      if (subsegment) {
        subsegment.close();
      }
    }
  });
};
