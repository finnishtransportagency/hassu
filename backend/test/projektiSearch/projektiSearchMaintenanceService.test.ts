import { describe, it } from "mocha";
import {
  OpenSearchClient,
  openSearchClientIlmoitustauluSyote,
  openSearchClientJulkinen,
  openSearchClientYllapito,
} from "../../src/projektiSearch/openSearchClient";
import * as sinon from "sinon";
import { ProjektiSearchMaintenanceService } from "../../src/projektiSearch/projektiSearchMaintenanceService";

const sandbox = sinon.createSandbox();

const { expect } = require("chai");

describe("ProjektiSearchMaintenanceService", () => {
  let yllapitoStub: sinon.SinonStubbedInstance<OpenSearchClient>;
  let suomiStub: sinon.SinonStubbedInstance<OpenSearchClient>;
  let ruotsiStub: sinon.SinonStubbedInstance<OpenSearchClient>;
  let saameStub: sinon.SinonStubbedInstance<OpenSearchClient>;
  let ilmoitustauluStub: sinon.SinonStubbedInstance<OpenSearchClient>;

  before(() => {
    yllapitoStub = sandbox.stub(openSearchClientYllapito);
    suomiStub = sandbox.stub(openSearchClientJulkinen["SUOMI"]);
    ruotsiStub = sandbox.stub(openSearchClientJulkinen["RUOTSI"]);
    saameStub = sandbox.stub(openSearchClientJulkinen["SAAME"]);
    ilmoitustauluStub = sandbox.stub(openSearchClientIlmoitustauluSyote);
  });
  afterEach(() => {
    sandbox.reset();
  });
  after(() => {
    sandbox.restore();
  });

  it("should delete and create index", async () => {
    await new ProjektiSearchMaintenanceService().deleteIndex();
    [yllapitoStub, suomiStub, ruotsiStub, saameStub].map((clientStub) => {
      [
        clientStub.put,
        clientStub.delete,
        clientStub.putSettings,
        clientStub.putMapping,
        ilmoitustauluStub.delete,
      ].forEach((operationStub) => expect(operationStub.calledOnce));
    });
  });
});
