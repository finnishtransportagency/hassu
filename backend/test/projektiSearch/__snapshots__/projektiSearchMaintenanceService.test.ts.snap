// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjektiSearchMaintenanceService should delete and create index 1`] = `
Object {
  "put": Array [
    Object {
      "args": Array [
        "projekti",
        "",
      ],
    },
  ],
}
`;

exports[`ProjektiSearchMaintenanceService should delete and create index 2`] = `
Object {
  "delete": Array [],
}
`;

exports[`ProjektiSearchMaintenanceService should delete and create index 3`] = `
Object {
  "putSettings": Array [
    Object {
      "args": Array [
        "projekti",
      ],
    },
  ],
}
`;

exports[`ProjektiSearchMaintenanceService should delete and create index 4`] = `
Object {
  "putMapping": Array [
    Object {
      "args": Array [
        "projekti",
      ],
    },
  ],
}
`;
